### ISSUE TITLE
connectivity_key error - tutorial not working 

### ISSUE THREAD

**Issue by inofechm** [wanglab-broad/spin#4]
Hello I am interesting in trying to use SPIN, but the code isn't working on my end both with own data and with the tutorial:
I have installed SPIN into a completely new virtual environment.

Here is the error (seems to be something with the adata structure). Not sure if this is an issue with a newer version of scanpy or something else.

```
adata_marmoset = sc.read('data/spin_paper/test/marmoset.h5ad', backup_url='https://zenodo.org/record/8092024/files/marmoset.h5ad?download=1')

adata_marmoset = spin(adata_marmoset,  n_pcs=20, resolution=0.3 )

2024-04-19 09:17:19,269 - SPIN - INFO - Smoothing

2024-04-19 09:17:20,658 - SPIN - INFO - Performing PCA
2024-04-19 09:17:21,690 - SPIN - INFO - Finding latent neighbors
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
2024-04-19 09:18:00,296 - SPIN - INFO - Leiden clustering
/Users/inofechmozes/Documents/venvs/spin/lib/python3.9/site-packages/spin/spin.py:295: FutureWarning: In the future, the default backend for leiden will be igraph instead of leidenalg.

 To achieve the future defaults please pass: flavor="igraph" and n_iterations=2.  directed must also be False to work with igraph's implementation.
  sc.tl.leiden(
2024-04-19 09:18:32,336 - SPIN - INFO - Performing UMAP
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/inofechmozes/Documents/venvs/spin/lib/python3.9/site-packages/spin/spin.py", line 113, in spin
    adata = _cluster(
  File "/Users/inofechmozes/Documents/venvs/spin/lib/python3.9/site-packages/spin/spin.py", line 304, in _cluster
    umap = sc.tl.umap(
  File "/Users/inofechmozes/Documents/venvs/spin/lib/python3.9/site-packages/legacy_api_wrap/__init__.py", line 80, in fn_compatible
    return fn(*args_all, **kw)
  File "/Users/inofechmozes/Documents/venvs/spin/lib/python3.9/site-packages/scanpy/tools/_umap.py", line 160, in umap
    neighbors = NeighborsView(adata, neighbors_key)
  File "/Users/inofechmozes/Documents/venvs/spin/lib/python3.9/site-packages/scanpy/_utils/__init__.py", line 1019, in __init__
    self._conns_key = self._neighbors_dict["connectivities_key"]
KeyError: 'connectivities_key'
```
Let me know how I can fix this if possible.

Thank you.

**Comment by kmaherx** [comment]
Hi!

Sorry for the delay, missed the notification of your issue.

After a bit of a dive, I think the core of this issue lies in the new version of Scanpy and its implementation of UMAP.
While I feel this problem is a bit beyond my depth, I found a workaround that seems to fix the problem on my end.
In short, I found that Leiden seems to be changing attributes of the AnnData object such that UMAP can no longer find information regarding neighbors. While I can't trace it back to some core issue at the moment, I did find that simply swapping the order in which Leiden and UMAP are performed appears to fix the issue. I made a branch called `swap_leiden_umap_order` that reflects these changes.

Give it a try by installing from the following link: `pip install git+https://github.com/wanglab-broad/spin@swap_leiden_umap_order`
If it works, I'll likely just merge the branch to `main` and call it done.

Let me know what you think!

Best,
Kamal
